sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctAttributesVisitor.scala

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeSet, ExpressionSet, NamedExpression}
+import org.apache.spark.sql.catalyst.plans.{LeftAnti, LeftSemi}
+
+/**
+ * A visitor pattern for traversing a [[LogicalPlan]] tree and propagate the distinct attributes.
+ */
+object DistinctAttributesVisitor extends LogicalPlanVisitor[Seq[AttributeSet]] {
```

Comment:
nit: `DistinctKeyVisitor` seems a simpler and more general name

Reply:
BTW, I don't think the order of `ExpressionSet` matters, and `Set[ExpressionSet]` is better

Reply:
and shall we make it a trait so that `LogicalPlan` can extend it directly? then we don't need `LogicalPlanDistinctKeys`. I think it's simpler if we only have one visitor implementation for distinct keys, which should be true.

Reply:
+ 1 to make it a trait.